### PR TITLE
Added a dedicated option type for "QueryExpression"

### PR DESCRIPTION
**Pull request by huppeje**
https://coveord.atlassian.net/browse/JSUI-2172

Added a dedicated option type for "QueryExpression"

### CODE REVIEW

File: src/ui/Base/ComponentOptions.ts [hunk 1]
```diff
@@ -742,6 +759,13 @@ export class ComponentOptions {
     return field;
   }
 
+  static loadQueryExpressionOption(element: HTMLElement, name: string, option: IComponentOptionsOption<any>): string {
+    const queryExpression = ComponentOptions.loadStringOption(element, name, option);
+    // TODO REVIEW : Will we want to make a specific loading process?
```
[the remark this answers is not in the document]
Comment (huppeje) [reply src/ui/Base/ComponentOptions.ts]: 👍

File: src/ui/Base/ComponentOptions.ts [hunk 2]
```diff
@@ -17,6 +17,11 @@ import { Template } from '../Templates/Template';
  */
 export interface IFieldOption extends String {}
 
+/**
+ * TODO : Documentation
```
[the remark this answers is not in the document]
Comment (huppeje) [reply src/ui/Base/ComponentOptions.ts]: Yes sure! 👍

File: src/ui/Base/ComponentOptions.ts [hunk 3]
```diff
@@ -495,6 +501,17 @@ export class ComponentOptions {
     return ComponentOptions.buildOption<string[]>(ComponentOptionsType.FIELDS, ComponentOptions.loadFieldsOption, optionArgs);
   }
 
+  /**
+   * TODO : documentation
```
[the remark this answers is not in the document]
Comment (huppeje) [reply src/ui/Base/ComponentOptions.ts]: Yes sure! 👍

**Comment by huppeje** [comment]
Yes sure! Here are the screenshots took from the updated documentation website : 

capture d ecran 2018-07-20 a 16 27 21


capture d ecran 2018-07-20 a 16 29 20

**Comment by olamothe** [comment]
For the screenshot, I mean for example, the "expression" option on the tab component, what does it look like now ?

**Comment by huppeje** [comment]
Ooo I see! There it is : 

capture d ecran 2018-07-23 a 11 59 15
 

It is still keeping the type 'string', is it what we want?

File: src/ui/Base/ComponentOptions.ts [hunk 4]
```diff
@@ -495,6 +503,22 @@ export class ComponentOptions {
...
   }
 
+  /**
+   * Builds a query expression option.
+   *
+   * A query expression option can be any arbitrary string in the markup.
+   *
+   * **Markup Example:**
+   *
+   * > `data-foo="@bar==baz"`
```
[the remark this answers is not in the document]
Comment (huppeje) [reply src/ui/Base/ComponentOptions.ts]: foo, bar & baz are placeholder names for variables. I based this markup example on the buildFieldsOption's documentation (see line 491 of this file) : 

```
   * Builds an array of fields option.
   *
   * As with all options that expect an array, you should use commas to delimit the different values.
   *
   * **Markup Example:**
   *
   * > `data-foo="@bar,@baz"`
```

[the remark this answers is not in the document]
Comment (amoreauCoveo) [reply src/ui/Base/ComponentOptions.ts]: I know they're placeholders, but what I'm saying is that right now it's confusing as to what is accepted and what isn't. If anything, you say it's an arbitrary string, so why not just use `foo`? (Note: I don't think we should just write `foo`).

I guess what I'm trying to say is that I don't understand what this component does enough to be able to fix it.

File: src/ui/Base/ComponentOptions.ts [hunk 5]
```diff
@@ -495,6 +503,22 @@ export class ComponentOptions {
     return ComponentOptions.buildOption<string[]>(ComponentOptionsType.FIELDS, ComponentOptions.loadFieldsOption, optionArgs);
   }
 
+  /**
+   * Builds a query expression option.
+   *
+   * A query expression option can be any arbitrary string in the markup.
```
[the remark this answers is not in the document]
Comment (olamothe) [reply src/ui/Base/ComponentOptions.ts]: Indeed, we can add a link to the online documentation that describe the Coveo query syntax.

http://www.coveo.com/go?dest=cloudhelp&lcid=9&context=357

[the remark this answers is not in the document]
Comment (huppeje) [reply src/ui/Base/ComponentOptions.ts]: Technically, any string can be added in the markup. Indeed, a wrong expression will not work. You are right, we probably should specify that the expression should follow the Coveo Query Syntax.

Adding this line under the line 509 would work? : 

```
* Whether to have a valid expression in the query expression option see
* [Coveo Query Syntax Reference](http://www.coveo.com/go?dest=adminhelp70&lcid=9&context=10005).
```

[the remark this answers is not in the document]
Comment (huppeje) [reply src/ui/Base/ComponentOptions.ts]: I will make the change! 👍

File: src/ui/Base/ComponentOptions.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (huppeje) [reply src/ui/Base/ComponentOptions.ts]: The `QueryExpressionOption` is a tool that allows the user to build a valid query expression. 

`data-foo=` : is the attribute where the query expression will be stored. 
`"@bar==baz"` : is a simple query expression.

We will remove the part ```A query expression option can be any arbitrary string in the markup.```  it should be less confusing

[the remark this answers is not in the document]
Comment (amoreauCoveo) [reply src/ui/Base/ComponentOptions.ts]: I did not understand and now I do. It's fine :P

**Pull request merged**